### PR TITLE
Outbound shipment auto allocation

### LINKED ISSUE #2227: When issuing stock, alert if a line isn't allocated
## What went wrong? 😲
When you issue stock either on an outbound shipment, or a prescription, you don't get an alert if you click OK on the modal without actually allocating stock. 

This happens both if you forget to allocate stock, and if there is an expired stockline (so it doesn't get issued). I think this will leave a lot of customers confused (and I am often confused too!) 

## Expected behaviour 🤔
I would like an alert if I have not allocated stock when closing the Issue Item modal. 

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Create an outbound shipment or prescription 
2. Click "Add item"
3. Select an item, and then (optional, but my general workflow) enter in a quantity
4. Now click OK - note that the line has not been added, but there is no alert. 

This also happens if you DO allocate but the lines are expired 

## Your environment 🌱

 - Version: 1.3.0
 - Platform:
   - all

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #2227, #2266, #2264, #1983, #2255

# 👩🏻‍💻 What does this PR do? 
 
See #2261 for the discussion and requested changes. Reinstates the auto allocation which was removed in #1189

- Allow saving of lines with 0 quantity
- Show a warning if you click OK when the quantity is 0; this asks you to click OK again to confirm
- Reinstate the auto allocation, this happens when the issue quantity is changed or the pack size is changed
- Prune 0 quantity lines when changing status from NEW
- Show a warning when changing status
- Fix the strange warning issue - where the notification shows incorrectly
- Show the warning when the pack size prevents the requested quantity being issued
- Show a warning when a placeholder has been allocated due to insufficient stock

Screenshot 2023-09-27 at 11 39 12 AM

Screenshot 2023-09-27 at 11 38 44 AM

Screenshot 2023-09-27 at 10 31 44 PM

Screenshot 2023-09-27 at 7 10 23 PM

# 🧪 How has/should this change been tested? 

- Create a new outbound shipment and try to add an item with 0 quantity
- Update an existing line to 0  and check the messaging
- Try the auto allocation
- Try the two way updates - editing a line and checking that the issue quantity updates
- Change status and check that lines are pruned

## 📃 Documentation

- [ ]  Auto allocation
- [ ] UI changes

### CODE REVIEW

**Comment by mark-prins** [comment]
Just draft at this stage, to get feedback. Haven't done prescriptions yet.

**Comment by roxy-dao** [comment]
Looks good to me so far ~, thanks Mark~

imo, the `placeholder-allocated` isn't clear enough... and there are these two issues to address this #2255 and #1983 🤔

**Comment by CarlosNZ** [comment]
Cool, looking good. Here's just a few observations (in no particular order) from my initial testing:

- Maybe the auto-allocation should only happen when losing focus rather than on every keystroke of the quantity input? I guess your reasoning is that you probably want to go straight from quantity entry to "OK" and you should see the allocations before clicking OK? I guess that makes sense, but perhaps a short debounce then on the input? (It's particularly noticeable when you add a quantity that causes the "Due to the pack sizes available a total quantity ..." message to appear -- when typing a number, you get the warning after the first keystroke which then disappears, but it shouldn't really appear until you've finished typing you're intended number -- although maybe this is a separate issue?)
- Shouldn't the "OK & Next" button behave the same as the "OK" button in terms of the zero-quantity warning? Currently it just goes to the next with no feedback.
- If we add a quantity, but then go back in to that line and change the quantity to 0, we get the warning and the 0 allocation display, but when clicking "OK" the line in the table hasn't been updated (to 0) -- still shows previous quantity, with "black" formatting.
- This is pretty nitpicky, but I feel like the top/bottom margins around the warning boxes could be a little wider? 
Screenshot 2023-10-03 at 11 54 04 AM  
Maybe somewhere between where it is currently and the margin below the "Issue" line?
- There's an error that happens in some situations, though I'm not 100% clear what it is. Here's how I can reproduce it on my setup:
  - Open a line and add a quantity higher than available. You'll get the "Not enough stock" warning. 
  - Close the item
  - Open another item and input a quantity
  - Even when this quantity is valid, the "OK" button won't work (and sometimes an error snackbar appears)
  - I suspect the warning from the previous item is still "active" somehow, so we probably need to explicitly clear it when closing the modal

I'll have a closer look at the code once this is no longer in Draft. Working nicely though and I like the warning when 0 solution.

**Comment by mark-prins** [comment]
Thanks for the comments @roxy-dao & @CarlosNZ : super helpful!!

> Maybe the auto-allocation should only happen when losing focus..

I did actually try both the `onBlur` and a debounce - and missed having the immediate feedback between what I type and the changes that happen with the stock line allocation.

Even having a debounce made it harder to get an intuitive sense that what I've entered related to the stock lines. Having the notifications pop in was annoying but I thought the trade off worth it. I'll try adding a small debounce to the notification display and see how that feels 🙏

**Comment by mark-prins** [comment]
Updated with review feedback and have added to prescriptions.
After implementing the more flexible `Alert` component, I've removed `InfoPanel` and replaced usages, and added a story.

The prescriptions now allow adding of `0` quantity rows, if the prescription is `New` only. If you add when `Picked` the rows are removed automatically. The rows are also pruned when you changed to `Verified`. I thought that this is the best compromise for the upcoming change to always have prescriptions `Picked` when items are added.. though of course things become a little weird. I don't expect that anyone would add a 0 quantity prescription line, so this is a very niche scenario anyway. 

Couldn't reproduce your last point @CarlosNZ and so I don't know if that's resolved or data specific. Can you raise a new issue if that's a valid problem still.

Screenshot 2023-10-04 at 5 42 50 PM
Screenshot 2023-10-04 at 10 14 42 PM
Screenshot 2023-10-05 at 8 19 22 AM
Screenshot 2023-10-05 at 8 21 13 AM

**Comment by CarlosNZ** [comment]
Hi @mark-prins -- I'm getting server compilation errors on this branch now (which are not on develop).

The console produces screeds of text, but here's the last bit if that helps:
Screenshot 2023-10-05 at 1 16 40 PM

**Comment by roxy-dao** [comment]
> Hi @mark-prins -- I'm getting server compilation errors on this branch now (which are not on develop).
> 
try `cargo clean`

**Comment by CarlosNZ** [comment]
> > Hi @mark-prins -- I'm getting server compilation errors on this branch now (which are not on develop).
> 
> try `cargo clean`

Thanks -- sorted :)

**Review by CarlosNZ — COMMENTED** [review]
Looking good, thanks. I've made one little comment about the message persisting on "OK and Next".

Couple of overall thoughts:
- I think I preferred the previous iteration when the notification boxes weren't full width. They seem to have too much empty space on the right now.
- It seems a bit odd to have the message "Some stock lines are expired and cannot be allocated" not appear till after you've allocated something. This info is known from the get-go, so it feels like that message should already be there when you open the modal. Would that be hard to do? (The other messages make sense when they appear)
- I can't reproduce the bug I mentioned in the last review any more, so whatever you've done seems to have fixed it 👍

**Comment by mark-prins** [comment]
> I think I preferred the previous iteration when the notification boxes weren't full width. They seem to have too much empty space on the right now.

oh. unintentional - will see what change did that.

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx [hunk 1]
```diff
@@ -159,6 +152,56 @@ export const OutboundLineEdit: React.FC<ItemDetailsModalProps> = ({
...
+      setShowZeroQuantityConfirmation(true);
+      return;
+    }
+
+    try {
+      await onSave();
+      setIsDirty(false);
+      if (!!placeholder) {
+        const infoSnack = info(t('message.placeholder-line'));
+        infoSnack();
```
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:169]: no reason that I can see other than making it clearer. I've had people review and insist that it be split out, iirc that was just about clarity. In this case - the lines are pre-exisiting, simply moved from elsewhere in the file.

**Comment by mark-prins** [comment]
> It seems a bit odd to have the message...

I prefer having it tied to the auto allocation. I don't want to have an alert message saying that there's on hold / expired stock which shows all the time, particularly when the individual lines are already showing this (red font for the expiry date and the 'on hold' column). The intention of the message is to explain why the allocation behaved how it did, and I'd rather not clutter the UI with an additional message.

**Review by CarlosNZ — APPROVED** [review]
Okay, LGTM.

**Pull request merged**